File: src/com/cattailsw/nanidroid/install/NarStagedSourceCopyResult.java
package com.cattailsw.nanidroid.install;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/** Immutable internal result that exposes only the staged capability. */
final class NarStagedSourceCopyResult {
    private final NarStagedSource source;
    private final NarStagedSourceCopyError error;
    private final String detail;
    private final List<NarStagedSourceCopyError> cleanupErrors;

    private NarStagedSourceCopyResult(
            NarStagedSource source,
            NarStagedSourceCopyError error,
            String detail,
            List<NarStagedSourceCopyError> cleanupErrors) {
        this.source = source;
        this.error = error;
        this.detail = detail;
        this.cleanupErrors = Collections.unmodifiableList(
                new ArrayList<NarStagedSourceCopyError>(
                        cleanupErrors));
    }

    static NarStagedSourceCopyResult success(
            NarStagedSource source) {
        return new NarStagedSourceCopyResult(
                source,
                null,
                "",
                Collections
                        .<NarStagedSourceCopyError>emptyList());
    }

    static NarStagedSourceCopyResult failure(
            NarStagedSourceCopyError error,
            String detail,
            List<NarStagedSourceCopyError> cleanupErrors) {
        return new NarStagedSourceCopyResult(
                null, error, detail, cleanupErrors);
    }

    boolean isSuccess() {
        return source != null;
    }

    NarStagedSource getSource() {
        return source;
    }

    NarStagedSourceCopyError getError() {
        return error;
    }

    String getDetail() {
        return detail;
    }

    List<NarStagedSourceCopyError> getCleanupErrors() {
        return cleanupErrors;
    }
}

File: src/com/cattailsw/nanidroid/install/NarStagedSourceCopyResult.kt
package com.cattailsw.nanidroid.install

import java.util.ArrayList
import java.util.Collections

/** Immutable internal result that exposes only the staged capability. */
internal class NarStagedSourceCopyResult private constructor(
    private val source: NarStagedSource?,
    private val error: NarStagedSourceCopyError?,
    private val detail: String?,
    cleanupErrors: List<NarStagedSourceCopyError>
) {
    private val cleanupErrors = Collections.unmodifiableList(
        ArrayList(cleanupErrors)
    )

    companion object {
        @JvmStatic
        fun success(source: NarStagedSource?): NarStagedSourceCopyResult =
            NarStagedSourceCopyResult(source, null, "", emptyList())

        @JvmStatic
        fun failure(
            error: NarStagedSourceCopyError?,
            detail: String?,
            cleanupErrors: List<NarStagedSourceCopyError>
        ): NarStagedSourceCopyResult =
            NarStagedSourceCopyResult(null, error, detail, cleanupErrors)
    }

    fun isSuccess(): Boolean = source != null

    fun getSource(): NarStagedSource? = source

    fun getError(): NarStagedSourceCopyError? = error

    fun getDetail(): String? = detail

    fun getCleanupErrors(): List<NarStagedSourceCopyError> = cleanupErrors
}
